gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/experimental/internal/configureNativeCompilerCaching.kt

```diff
@@ -0,0 +1,114 @@
+private const val COMPOSE_NATIVE_MANAGE_CACHE_KIND = "compose.kotlin.native.manageCacheKind"
+private const val NONE_VALUE = "none"
+
+internal fun Project.configureNativeCompilerCaching() {
+    if (findProperty(COMPOSE_NATIVE_MANAGE_CACHE_KIND) == "false") return
+
+    plugins.withId(KOTLIN_MPP_PLUGIN_ID) {
+        val kotlinVersion = kotlinVersionNumbers(this)
+        mppExt.targets.configureEachWithType<KotlinNativeTarget> {
+            assertCacheKindUserValueIsNotNone()
+            configureTargetCompilerCache(kotlinVersion)
+        }
+    }
+}
+
+private fun KotlinNativeTarget.assertCacheKindUserValueIsNotNone() {
+    // To determine cache kind KGP checks kotlin.native.cacheKind.<PRESET_NAME> first, then kotlin.native.cacheKind
+    // For each property it tries to read Project.property, then checks local.properties
+    // See https://github.com/JetBrains/kotlin/blob/d4d30dcfcf1afb083f09279c6f1ba05031efeabb/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/PropertiesProvider.kt#L416
+    val cacheKindProperties = listOf(targetCacheKindPropertyName, PROJECT_CACHE_KIND_PROPERTY_NAME)
+    val propertyProviders = listOf(
+        KGPPropertyProvider.GradleProperties(project),
+        KGPPropertyProvider.LocalProperties(project)
+    )
+
+    for (cacheKindProperty in cacheKindProperties) {
+        for (provider in propertyProviders) {
+            val value = provider.valueOrNull(cacheKindProperty)
+            if (value != null) {
+                if (value.equals(NONE_VALUE, ignoreCase = true)) {
+                    project.logger.warn(cacheKindPropertyWarningMessage(cacheKindProperty, provider))
+                }
+                return
+            }
+        }
+    }
+}
+
+private fun cacheKindPropertyWarningMessage(
+    cacheKindProperty: String,
+    provider: KGPPropertyProvider
+) = """
+    |Warning: '$cacheKindProperty' is explicitly set to `none`.
+    |This option significantly slows the Kotlin/Native compiler.
+    |Compose Multiplatform Gradle plugin can set this property automatically,
+    |when it is necessary.
+    |  * Recommended action: remove explicit '$cacheKindProperty=$NONE_VALUE' from ${provider.location}. 
+    |  * Alternative action: if you are sure you need '$cacheKindProperty=$NONE_VALUE', disable
+    |this warning by adding '$COMPOSE_NATIVE_MANAGE_CACHE_KIND=false' to your 'gradle.properties'.
+""".trimMargin()
+
+private fun KotlinNativeTarget.configureTargetCompilerCache(kotlinVersion: KotlinVersionNumbers) {
+    val (majorVer, minorVer, patchVer) = kotlinVersion
+    // See comments in https://youtrack.jetbrains.com/issue/KT-57329
+    // When:
+    //    KotlinVersion < 1.9.0,          => set cacheKind=none
+    //    1.9.0 <= KotlinVersion < 1.9.20 => add -Xlazy-ir-for-caches=disable
+    //    KotlinVersion >= 1.9.20         => do nothing, everything should work as is
+    when {
+        majorVer < 1 || majorVer == 1 && minorVer < 9 -> {
```

Comment:
Can we just declare `KotlinVersionNumbers.compareTo` operator?

Reply:
Good idea, thanks

Reply:
BTW can we reuse [KotlinVersion](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-kotlin-version/) class?